net/peer_tracker.go

```diff
@@ -18,32 +22,98 @@ var logPeerTracker = logging.Logger("peer-tracker")
 type PeerTracker struct {
 	// mu protects peers
 	mu sync.RWMutex
-	// peers maps stringified peer.IDs to info about their chains
-	peers map[string]*types.ChainInfo
+	// peers maps peer.IDs to info about their chains
+	peers    map[peer.ID]*types.ChainInfo
+	trusted  map[peer.ID]bool
+	updateFn updatePeerFn
 }
+type updatePeerFn func(ctx context.Context, p peer.ID) (*types.ChainInfo, error)
 
 // NewPeerTracker creates a peer tracker.
-func NewPeerTracker() *PeerTracker {
+func NewPeerTracker(trust ...peer.ID) *PeerTracker {
+	trustedSet := make(map[peer.ID]bool)
+	for _, t := range trust {
+		trustedSet[t] = true
+	}
 	return &PeerTracker{
-		peers: make(map[string]*types.ChainInfo),
+		peers:   make(map[peer.ID]*types.ChainInfo),
+		trusted: trustedSet,
+	}
+}
+
+// SetUpdateFn sets the update function `f` on the peer tracker. This function is a prerequisite
+// to the UpdateTrusted logic.
+func (tracker *PeerTracker) SetUpdateFn(f updatePeerFn) {
+	tracker.updateFn = f
+}
+
+// SelectHead returns the chain info with the greatest height being tracked. If `trust` is true then
+// only the trusted tracked nodes will be considered. An error is returned if no peers are in the tracker.
+func (tracker *PeerTracker) SelectHead(trust bool) (types.ChainInfo, error) {
```

Comment:
Moving forward I see this method feeding the syncer during the initial sync, usage would be something like:
```go
for NotCaughtUp() {
  peertracerk.UpdatePeers()
  head := peertracker.SelectHead()
  syncer.HandleNewTipSet(head)
}
```